### PR TITLE
CHARTER: Add WG proposal and templates

### PR DESCRIPTION
**Pull request by swinslow**
(cc @caniszczyk)

This is a revised version of the OCI Working Group Draft Proposal for consideration by the TOB.

I've reworked the draft proposal somewhat, moving some details (such as lists of specific info that the TOB would require for review) into a separate `WG-INFO.md` document. That way those pieces of information could be modified by the TOB in its discretion, without requiring a subsequent formal charter amendment.

I've also made some proposed edits to the language of what remains in the charter portion, I'd encourage folks to please review closely. A minor note, I've also shifted the Working Group content into section 6(p) rather than as a new section 7, to avoid having to do lots of renumbering for the rest of the charter.

Finally, I'm including a `WG-TEMPLATE.md` document which could be used by the TOB as a way to document the scope / work product and ways of working for each new WG, but the TOB could of course modify or omit this altogether if it doesn't like this approach.

If the TOB does elect to proceed with adopting this, please note the procedure in section 12(a) of the OCI Charter for adopting charter amendments, including the required period of notice to members.

Signed-off-by: Steve Winslow <steve@swinslow.net>

### CODE REVIEW

**Comment by caniszczyk** [comment]
RFC @opencontainers/tob

File: CHARTER.md
Comment (vsoch) [inline CHARTER.md:343]: The implication here is that someone that submits a draft that is approved becomes the maintainers (which is a commitment, so maybe it should be stated).

File: CHARTER.md
Comment (vsoch) [inline CHARTER.md:352]: "meaningful progress" could that be better defined? And inactive is easier, e.g., "no discussion or updates for X amount of time." Reading this I'd be afraid that my group would be disbanded pretty subjectively.

File: WG-TEMPLATE.md
Comment (vsoch) [inline WG-TEMPLATE.md:30]:
```suggestion
  * The TOB is establishing the WG as an OCI Working Group, pursuant to [section 6(p)](https://github.com/opencontainers/tob/blob/master/CHARTER.md#6-technical-oversight-board-tob) of the OCI Charter.
```
And note that section p is not in the current document, it's defined in this PR.

File: WG-TEMPLATE.md
Comment (vsoch) [inline WG-TEMPLATE.md:45]: For these links to sections, I'm wondering if they should be smaller headings so we can link directly to them? if this is of interest I can PR to do this (and update the links here) after this is merged.

File: WG-TEMPLATE.md
Comment (dmcgowan) [inline WG-TEMPLATE.md:31]: The charter section uses "owners" as term to represent this role. The owners of a working group are approved as part of the working group. I'm not sure I understand what "appoint" means in that context. Likewise, for making changes to a working group after it is approved, we should define the process as the owners request a change (in Amendments most likely). Any change requires the same level of approval as the initial proposal. It doesn't make sense to explicitly say that the TOB would remove or appoint new maintainers(/owners), the TOB would be responsible for either disbanding the working group or approving changes proposed by the owners, which could include changes to the owner or sponsor roles.

File: CHARTER.md
Comment (dmcgowan) [reply CHARTER.md:352]: These processes are always going to be subjective though with phrases like "in its discretion" or "meaningful". The idea behind any two-thirds vote is that it is the majority opinion. Whether the majority opinion is the right opinion is a different topic :smile:

File: CHARTER.md
Comment (dmcgowan) [reply CHARTER.md:343]: That was not actually the point and I think could be called out better. The original idea is that the working group would have owners, and when a draft is approved and becomes a specification, the set of maintainers is voted on at that time as well. Becoming a maintainer is a commitment, but there should be no requirement that owners are also the maintainers. Owners are responsible for finding maintainers though before a draft could be approved as a new specification/project. I commented below about disambiguating owners and maintainers.

File: CHARTER.md
Comment (vsoch) [reply CHARTER.md:320]: These would be great to have! It would clear up any confusion that a working group is a specific project / task as opposed to a community.

File: CHARTER.md
Comment (swinslow) [reply CHARTER.md:336]: Thanks @vsoch -- yes, I think this formal process in the charter is focused more on the stage at which a proposed group is far enough along to do a formal proposal to start a Working Group. For things that are more like experiments / initial testing, I think OCI could decide to create an "-experiments" or "-labs" or similar org, but I'm not sure that's something that needs to be explicitly called out in the charter.

File: CHARTER.md
Comment (swinslow) [reply CHARTER.md:337]: Hmmm... I was thinking that this read more as saying that this would happen at the end of the drafting process (when the WG "determines that it has achieved its objectives"). At least that was my intent! Do you have a suggestion for a clearer way to phrase this?

File: CHARTER.md
Comment (swinslow) [reply CHARTER.md:348]: @vsoch and @dmcgowan, good point. In the initial draft I hadn't fully appreciated some of the intended distinctions between owners and maintainers. In the follow-up commit I've rephrased some of the details here in the charter and in the WG governance documents. Take a look and let me know if you are good with the changes, or if there are other specific changes you'd like to see.

File: CHARTER.md
Comment (swinslow) [reply CHARTER.md]: @vsoch and @samuelkarp Great suggestions -- in the first draft, I'd been treating "OCI Projects" as if it was a distinct and separate thing from OCI Specifications. But looking at the defined terms earlier in the charter, it looks like "OCI Projects" encapsulates OCI Specifications as well as implementations like runc.

In the updated commit, I've tried to clarify this by referring to e.g. "OCI Specifications and other OCI Projects", and added some of the clarifying examples that @samuelkarp suggested.

File: CHARTER.md
Comment (swinslow) [reply CHARTER.md]: I'd tend to agree with @dmcgowan here. The "meaningful progress" is given as an example, but not a limiting one, so ultimately it's the TOB's determination (by a supermajority 2/3 vote) whether to disband.

File: WG-TEMPLATE.md
Comment (swinslow) [reply WG-TEMPLATE.md]: Thank you! I've incorporated this in a couple of places in the follow-up commit.

File: WG-TEMPLATE.md
Comment (swinslow) [reply WG-TEMPLATE.md]: That sounds like a good idea -- I'll defer to others but that sounds great to me.

File: WG-TEMPLATE.md
Comment (swinslow) [reply WG-TEMPLATE.md]: Thanks @dmcgowan -- all good points and thanks for clarifying. I've tried to update in the follow-up commit and clarify the differences between "owners" and "maintainers". Grateful if you can take a look and see if this works for you, or if there are further clarifications that should be added.

File: WG-INFO.md
Comment (swinslow) [reply WG-INFO.md]: Thanks! I've made updates along these lines in the follow-up commit to help clarify.

File: WG-TEMPLATE.md
Comment (SteveLasker) [inline WG-TEMPLATE.md:32]: Should we have a minimum of 2 or more owners? If something is to be run under OCI, seems it should be big enough, or interesting enough for multiple people to be engaged, to also assure the working group can scale to the needs of it being so visible under OCI.

File: CHARTER.md
Comment (vsoch) [reply CHARTER.md:358]: > The TOB should have the right to disband a working group if it's non-functional or inactive. 

Can there be a more structured process for this? E.g., let's say people are really busy and the group is dormant for a bit - and then they get an email "sorry disbanded!" Maybe a warning would come first with structured questions about status and plan, and then follow up check in two weeks to re-assess? We want the TOB to be able to govern and control non-functional or inactive groups, but having the process (with kindness) will come across more as a supportive structure than an unexpected dictator.

File: CHARTER.md
Comment (vsoch) [reply CHARTER.md]: Gotcha. Maybe when this is actual policy and this has happened a few times (if it even does) there could be examples of what is not meaningful progress. I think people tend to be busy on things (especially for extra obligations like a standards group) and would want to better understand what is not meaningful progress to avoid this event. Or give suggestions about positive behaviors, e.g.

- show up to weekly/monthly meetings to give a quick summary of progress
- have an active site that shows up to date meetings and activity

etc. etc. I think we would eventually just want this "scary case" of being disbanded and things to do to avoid it to be more concrete.

File: CHARTER.md
Comment (vsoch) [reply CHARTER.md:337]: Maybe add a step before it to prompt to define objectives, channels of communication and (if appropriate) timelines? E.g., some WG might just be working on a text document, and others might want a toy implementation. Then leading into this section would follow nicely.

File: CHARTER.md
Comment (vsoch) [reply CHARTER.md:336]: oh that would be cool :)

File: CHARTER.md
Comment (dlorenc) [reply CHARTER.md:317]: Major is pretty important here IMO. Working groups aren't the only way to make progress, specs already have maintainers and a change process.

File: WG-TEMPLATE.md
Comment (dlorenc) [reply WG-TEMPLATE.md:32]: +1 here - from multiple organizations would be great too.

File: CHARTER.md
Comment (dlorenc) [inline CHARTER.md:320]: This part is still a little unclear to me, when a WG is proposing changes to an existing specification that already has existing maintainers.

It seems like the following would be possible with this structure:
* a WG to form to make changes to an existing specification
* The WG decides on, and present these changes to the TOB for approval 
* The TOB ratifies these changes

all without any input from the existing maintainers.

Would it make sense to handle the case where changes are being proposed to existing specification(s) separately from a brand new specification?

File: CHARTER.md
Comment (dmcgowan) [reply CHARTER.md:317]: If the goals of the working group are completely in scope for an existing major release version the spec (for example changes to current 1.0 specs), then the working group is not necessary and should work with existing maintainers. A major revision of a spec (say image-spec 2.0) or a new spec should be considered very similar from a process perspective, no compatibility requirements with previous spec versions and separate set of maintainers who focus on the released major spec version.

What could be missing here is features which span multiple existing specs. There could be a working group which expects smaller changes in existing specs but not necessarily a separate released version as a result. In those cases the working group is expected to have maintainers of those specs as sponsors and the working group is dependent on those minor revisions getting released. As asked below, a working group is not intended to work around maintainers of a released spec version to push through a feature. If that needs to be more clear, then we can make it clear. Removing "major" here would necessitate that in my opinion.

File: CHARTER.md
Comment (dmcgowan) [reply CHARTER.md:320]: I tried to answer that above, existing maintainers could be sponsors and existing specs could be dependencies, but not the final result of the working group. I think in the case you mentioned where the working group covers existing specs, the TOB would vote to disband the working group after all the existing specs are updated or the working group cannot make progress because their changes were not accepted by the maintainers.

File: CHARTER.md
Comment (SteveLasker) [reply CHARTER.md:317]: The intent isn't to work around any process. Rather, there are questions that can only be answered through some iteration, incubation, and some sort of validation. The idea is to provide a means for innovation and let the result be reviewed for whether it should be incorporated into existing or new specs. At the beginning, the group may think it's simple to put in an existing spec. Or, it may be complex requiring a new spec. It's through the iteration, incubation and validations that a proposal is formalized and likely piviots in some direction. 
If the end result is known at the beginning, then we don't need a working group.

File: WG-TEMPLATE.md
Comment (dmcgowan) [reply WG-TEMPLATE.md:32]: Agreed, two or more from at least 2 organizations sounds good.

File: CHARTER.md
Comment (samuelkarp) [inline CHARTER.md:355]: nit: in `iv` above the term is "leaders" but her it's "owners"; we should use the same term in both places (whichever it is).

File: CHARTER.md
Comment (samuelkarp) [inline CHARTER.md:357]: nit: this is the first mention of a governance document for the working group.  Do we need to specify that ahead of time?

File: WG-INFO.md
Comment (samuelkarp) [inline WG-INFO.md:9]: nit: I think "former" might be less scary than "disbanded"

**Review by samuelkarp — APPROVED** [review]
Some minor nits for clarity, but in the spirit of actually moving forward:

LGTM!

**Review by OhArthits — APPROVED** [review]

File: CHARTER.md
Comment (vbatts) [inline CHARTER.md:346]: like [@swinslow comment](https://github.com/opencontainers/tob/pull/99/files?short_path=5825b09#r669885216), I wonder if these kinds of suffix can be grouped, or maybe have flexible language around _how_ we explicitly designate that they are not-released/experimental?

File: CHARTER.md
Comment (vbatts) [reply CHARTER.md:358]: @vsoch agreed. By the time it may come to a vote, I would imagine there to have been some communication. I would not expect a consensus on a vote that hasn't been discussed or surfaced to the affected project

**Review by vbatts — APPROVED** [review]
Like @samuelkarp, a couple of nits that we can address, but this sets the basics for us to work with now.
LGTM

**Review by dmcgowan — APPROVED** [review]
LGTM, remaining suggestions could be addressed in follow ups

**Review by fuweid — APPROVED** [review]
LGTM

**Review by estesp — COMMENTED** [review]
LGTM; looks like we are in good shape to get this finalized. I believe the actual implementation of a WG or two should expose any further clarifications that would be helpful and we are always free to make charter updates in the future if discrepancies are found in the implementation of the WG process.

**Review by estesp — APPROVED** [review]

**Comment by samuelkarp** [comment]
@opencontainers/tob Can we take a look at this?

@tych0 @jonjohnsonjr @SteveLasker @cyphar

**Review by SteveLasker — APPROVED** [review]
LGTM

**Comment by tych0** [comment]
One comment about adding some context, but other than that it looks reasonable to me. (I'm not par of the opencontainers/tob github team, so I didn't get the previous pings; anyone know how to fix that?)

**Comment by amye** [comment]
6 LGTMs (and @tych0 is maybe? an LGTM) out of 9, so I think we can move ahead here.

**Review by jonjohnsonjr — APPROVED** [review]
LGTM. I'd like to see the remaining feedback addressed, but we can address them in follow-up PRs.

I think once we start going through the motions, it will be easier to revise this with less theoretical feedback.

**Comment by tych0** [comment]
yep, if we want to add the links later that's fine. LGTM.

**Comment by cyphar** [comment]
Didn't we agree it would be re-drafted by LF legal in the last call (has that already happened -- I don't see a comment mentioning it?). I don't in principle have an issue with us merging it and ironing out any bugs in future revisions, but given that there isn't a drafting system for charter changes if we merge it now we're committing to this being the in-force charter until it gets re-drafted by legal.

**Comment by vbatts** [comment]
@caniszczyk @amye next steps?

**Comment by caniszczyk** [comment]
LGTMs 7/9 so this passes and can be merged

Sam: https://github.com/opencontainers/tob/pull/99#pullrequestreview-712286712
Tycho: https://github.com/opencontainers/tob/pull/99#issuecomment-893009491
Steve: https://github.com/opencontainers/tob/pull/99#pullrequestreview-720510636
Phil: https://github.com/opencontainers/tob/pull/99#pullrequestreview-717196321
Derek: https://github.com/opencontainers/tob/pull/99#pullrequestreview-716603670
Wei: https://github.com/opencontainers/tob/pull/99#pullrequestreview-717171279
Vincent: https://github.com/opencontainers/tob/pull/99#pullrequestreview-716297743

FYI @swinslow is LF Legal and helped draft the changes @cyphar 

We can iterate on future modifications but this should unblock the creation of focused WGs

**Pull request merged**
